rust-demangle.c

```diff
@@ -1147,30 +1152,173 @@ static void demangle_const_str_literal(struct rust_demangler *rdm) {
     PRINT("\"");
 }
 
-bool rust_demangle_with_callback(
-    const char *mangled, int flags,
-    void (*callback)(const char *data, size_t len, void *opaque), void *opaque
-) {
-    // Rust symbols always start with R, _R or __R.
-    if (mangled[0] == '_' && mangled[1] == 'R')
-        mangled += 2;
-    else if (mangled[0] == 'R')
-        // On Windows, dbghelp strips leading underscores, so we accept "R..."
-        // form too.
-        mangled += 1;
-    else if (mangled[0] == '_' && mangled[1] == '_' && mangled[2] == 'R')
-        // On OSX, symbols are prefixed with an extra _
-        mangled += 3;
-    else
+static bool is_rust_hash(struct rust_mangled_ident name) {
+    if (name.ascii[0] != 'h') {
         return false;
+    }
+    for (size_t i = 1; i < name.ascii_len; i++) {
+        if (!IS_DIGIT(name.ascii[i]) &&
+            !(name.ascii[i] >= 'a' && name.ascii[i] <= 'f')) {
+            return false;
+        }
+    }
+    return true;
+}
 
-    // Paths always start with uppercase characters.
-    if (!IS_UPPER(mangled[0]))
-        return false;
+static void print_legacy_ident(
+    struct rust_demangler *rdm, struct rust_mangled_ident ident
+) {
+    if (rdm->errored || rdm->skipping_printing)
+        return;
+
+    CHECK_OR(!ident.punycode, return);
+
+    if (ident.ascii[0] == '_' && ident.ascii[1] == '$') {
+        ident.ascii += 1;
+        ident.ascii_len -= 1;
+    }
+
+    while (1) {
+        if (ident.ascii_len == 0) {
+            break;
+        } else if (ident.ascii[0] == '.') {
+            if (ident.ascii_len >= 2 && ident.ascii[1] == '.') {
+                PRINT("::");
+                ident.ascii += 2;
+                ident.ascii_len -= 2;
+            } else {
+                PRINT(".");
+                ident.ascii += 1;
+                ident.ascii_len -= 1;
+            }
+        } else if (ident.ascii[0] == '$') {
+            const char *end_ptr =
+                (const char *)memchr(&ident.ascii[1], '$', ident.ascii_len - 1);
+            if (!end_ptr)
+                break;
+            const char *escape = &ident.ascii[1];
+            size_t escape_len = end_ptr - escape;
+
+            if (strncmp(escape, "SP", escape_len)  == 0) {
+                PRINT("@");
+            } else if (strncmp(escape, "BP", escape_len)  == 0) {
+                PRINT("*");
+            } else if (strncmp(escape, "RF", escape_len)  == 0) {
+                PRINT("&");
+            } else if (strncmp(escape, "LT", escape_len)  == 0) {
+                PRINT("<");
+            } else if (strncmp(escape, "GT", escape_len)  == 0) {
+                PRINT(">");
+            } else if (strncmp(escape, "LP", escape_len)  == 0) {
+                PRINT("(");
+            } else if (strncmp(escape, "RP", escape_len)  == 0) {
+                PRINT(")");
+            } else if (strncmp(escape, "C", escape_len)  == 0) {
+                PRINT(",");
+            } else {
```

Comment:
Hmm, this isn't a great idea without checking that `escape_len == 2` (which `strncmp` won't check for you).
If you want to have some fun with it, you could use a `switch(PACK(escape[0], escape[1]))` with `case PACK('S', 'P'):` where `#define PACK(a, b) ((uint16_t)(a) << 8) | (uint16_t)(b)`.